File: apps/ios/Shared/Views/Chat/Group/AddGroupRelayView.swift
//
//  AddGroupRelayView.swift
//  SimpleX (iOS)
//
//  Created by simplex on 29.04.2026.
//  Copyright © 2026 SimpleX Chat. All rights reserved.
//

import SwiftUI
import SimpleXChat

struct AddGroupRelayView: View {
    var groupInfo: GroupInfo
    var existingRelayIds: Set<Int64>
    var onRelayAdded: () -> Void
    @EnvironmentObject var theme: AppTheme
    @Environment(\.dismiss) var dismiss
    @State private var availableRelays: [(relayId: Int64, relay: UserChatRelay, operatorName: String?)] = []
    @State private var selectedRelayIds: Set<Int64> = []
    @State private var isLoading = true
    @State private var isAdding = false

    var body: some View {
        NavigationView {
            List {
                if isLoading {
                    Section {
                        ProgressView()
                            .frame(maxWidth: .infinity)
                    }
                } else if availableRelays.isEmpty {
                    Section {
                        Text("No available relays")
                            .foregroundColor(theme.colors.secondary)
                    }
                } else {
                    Section {
                        ForEach(availableRelays, id: \.relayId) { item in
                            relayCheckRow(item.relayId, item.relay, operatorName: item.operatorName)
                        }
                    }
                }
            }
            .modifier(ThemedBackground(grouped: true))
            .navigationTitle("Add relays")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
                ToolbarItem(placement: .confirmationAction) {
                    Button("Add") { addSelectedRelays() }
                        .disabled(selectedRelayIds.isEmpty || isAdding)
                }
            }
        }
        .task { await loadAvailableRelays() }
    }

    private func relayCheckRow(_ relayId: Int64, _ relay: UserChatRelay, operatorName: String?) -> some View {
        let selected = selectedRelayIds.contains(relayId)
        return Button {
            if selected {
                selectedRelayIds.remove(relayId)
            } else {
                selectedRelayIds.insert(relayId)
            }
        } label: {
            HStack {
                VStack(alignment: .leading) {
                    Text(chatRelayDisplayName(relay))
                        .foregroundColor(theme.colors.onBackground)
                        .lineLimit(1)
                    if let opName = operatorName {
                        Text(opName)
                            .font(.caption)
                            .foregroundColor(theme.colors.secondary)
                            .lineLimit(1)
                    }
                }
                Spacer()
                Image(systemName: selected ? "checkmark.circle.fill" : "circle")
                    .foregroundColor(selected ? theme.colors.primary : Color(uiColor: .tertiaryLabel).asAnotherColorFromSecondary(theme))
            }
        }
    }

    private func loadAvailableRelays() async {
        do {
            let servers = try await getUserServers()
            var relays: [(relayId: Int64, relay: UserChatRelay, operatorName: String?)] = []
            for op in servers {
                if let oper = op.operator, oper.enabled != true { continue }
                let opName: String? = op.operator?.operatorTag != nil ? op.operator?.tradeName : nil
                for relay in op.chatRelays {
                    if relay.enabled && !relay.deleted,
                       let relayId = relay.chatRelayId,
                       !existingRelayIds.contains(relayId) {
                        relays.append((relayId, relay, opName))
                    }
                }
            }
            await MainActor.run {
                availableRelays = relays
                isLoading = false
            }
        } catch {
            logger.error("loadAvailableRelays error: \(responseError(error))")
            await MainActor.run {
                isLoading = false
            }
        }
    }

    private func addSelectedRelays() {
        let relayIds = Array(selectedRelayIds)
        guard !relayIds.isEmpty else { return }
        isAdding = true
        Task {
            do {
                guard let result = try await apiAddGroupRelays(groupInfo.groupId, relayIds: relayIds) else {
                    await MainActor.run { isAdding = false }
                    return
                }
                await MainActor.run {
                    isAdding = false
                    switch result {
                    case let .added(gInfo, _, relays):
                        ChannelRelaysModel.shared.set(groupId: gInfo.groupId, groupRelays: relays)
                        onRelayAdded()
                        dismiss()
                    case let .addFailed(results):
                        let successIds = Set(results.filter { $0.relayError == nil }.compactMap { $0.relay.chatRelayId })
                        if !successIds.isEmpty {
                            selectedRelayIds.subtract(successIds)
                            availableRelays.removeAll { successIds.contains($0.relayId) }
                            onRelayAdded()
                        }
                        let errorLines = results.filter { $0.relayError != nil }
                            .map { "\(chatRelayDisplayName($0.relay)): \($0.relayError.map { connErrorText($0) } ?? "")" }
                        let successNames = results.filter { $0.relayError == nil }
                            .map { chatRelayDisplayName($0.relay) }
                        var msg = errorLines.joined(separator: "\n")
                        if !successNames.isEmpty {
                            msg += "\n" + String.localizedStringWithFormat(NSLocalizedString("Relays added: %@.", comment: "alert message"), successNames.joined(separator: ", "))
                        }
                        showAlert(
                            NSLocalizedString("Error adding relays", comment: "alert title"),
                            message: msg
                        )
                    }
                }
            } catch {
                await MainActor.run {
                    isAdding = false
                    showAlert(NSLocalizedString("Error adding relays", comment: "alert title"), message: responseError(error))
                }
            }
        }
    }
}

File: apps/multiplatform/common/src/commonMain/kotlin/chat/simplex/common/views/chat/group/AddGroupRelayView.kt
package chat.simplex.common.views.chat.group

import SectionBottomSpacer
import SectionCustomFooter
import SectionDividerSpaced
import SectionItemView
import SectionView
import androidx.compose.foundation.layout.*
import androidx.compose.material.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.simplex.common.model.*
import chat.simplex.common.platform.*
import chat.simplex.common.ui.theme.*
import chat.simplex.common.views.helpers.*
import chat.simplex.common.views.newchat.chatRelayDisplayName
import chat.simplex.common.views.usersettings.SettingsActionItem
import chat.simplex.res.MR
import dev.icerock.moko.resources.compose.painterResource
import kotlinx.coroutines.launch

data class AvailableRelay(
  val relayId: Long,
  val relay: UserChatRelay,
  val operatorName: String?
)

@Composable
fun AddGroupRelayView(
  groupInfo: GroupInfo,
  existingRelayIds: Set<Long>,
  onRelayAdded: () -> Unit,
  close: () -> Unit
) {
  var availableRelays by remember { mutableStateOf<List<AvailableRelay>>(emptyList()) }
  var selectedRelayIds by remember { mutableStateOf<Set<Long>>(emptySet()) }
  var isLoading by remember { mutableStateOf(true) }
  var isAdding by remember { mutableStateOf(false) }
  val scope = rememberCoroutineScope()

  BackHandler(onBack = close)

  LaunchedEffect(Unit) {
    try {
      val servers = ChatController.getUserServers(null)
      if (servers != null) {
        val relays = mutableListOf<AvailableRelay>()
        for (op in servers) {
          if (op.operator != null && op.operator.enabled != true) continue
          val opName: String? = if (op.operator?.operatorTag != null) op.operator.tradeName else null
          for (relay in op.chatRelays) {
            val relayId = relay.chatRelayId
            if (relay.enabled && !relay.deleted && relayId != null && relayId !in existingRelayIds) {
              relays.add(AvailableRelay(relayId, relay, opName))
            }
          }
        }
        availableRelays = relays
      }
    } catch (e: Exception) {
      Log.e(TAG, "loadAvailableRelays error: ${e.message}")
    }
    isLoading = false
  }

  AddGroupRelayLayout(
    availableRelays = availableRelays,
    selectedRelayIds = selectedRelayIds,
    isLoading = isLoading,
    isAdding = isAdding,
    onToggleRelay = { relayId ->
      selectedRelayIds = if (relayId in selectedRelayIds) selectedRelayIds - relayId else selectedRelayIds + relayId
    },
    onAddRelays = {
      val relayIds = selectedRelayIds.toList()
      if (relayIds.isEmpty()) return@AddGroupRelayLayout
      isAdding = true
      scope.launch {
        addSelectedRelays(groupInfo, relayIds, selectedRelayIds, availableRelays, onRelayAdded, close) { newSelectedIds, newAvailableRelays ->
          selectedRelayIds = newSelectedIds
          availableRelays = newAvailableRelays
          isAdding = false
        }
      }
    }
  )
}

@Composable
private fun AddGroupRelayLayout(
  availableRelays: List<AvailableRelay>,
  selectedRelayIds: Set<Long>,
  isLoading: Boolean,
  isAdding: Boolean,
  onToggleRelay: (Long) -> Unit,
  onAddRelays: () -> Unit
) {
  ColumnWithScrollBar {
    AppBarTitle(generalGetString(MR.strings.add_relays_title))

    if (isLoading) {
      Box(Modifier.fillMaxWidth().padding(vertical = DEFAULT_PADDING), contentAlignment = Alignment.Center) {
        CircularProgressIndicator()
      }
    } else if (availableRelays.isEmpty()) {
      SectionView {
        SectionItemView(padding = PaddingValues(horizontal = DEFAULT_PADDING)) {
          Text(
            generalGetString(MR.strings.no_available_relays),
            color = MaterialTheme.colors.secondary
          )
        }
      }
    } else {
      SectionView {
        AddRelaysButton(
          onClick = onAddRelays,
          disabled = selectedRelayIds.isEmpty() || isAdding
        )
      }
      SectionCustomFooter {
        val count = selectedRelayIds.size
        Text(
          if (count == 0) generalGetString(MR.strings.no_relays_selected)
          else String.format(generalGetString(MR.strings.num_relays_selected), count),
          color = MaterialTheme.colors.secondary,
          lineHeight = 18.sp,
          fontSize = 14.sp
        )
      }
      SectionDividerSpaced(maxTopPadding = true)
      SectionView(generalGetString(MR.strings.select_relays).uppercase()) {
        availableRelays.forEach { item ->
          val selected = item.relayId in selectedRelayIds
          SectionItemView(
            click = { onToggleRelay(item.relayId) },
            padding = PaddingValues(horizontal = DEFAULT_PADDING, vertical = 4.dp)
          ) {
            Column(Modifier.weight(1f)) {
              Text(
                chatRelayDisplayName(item.relay),
                maxLines = 1,
                color = MaterialTheme.colors.onBackground
              )
              if (item.operatorName != null) {
                Text(
                  item.operatorName,
                  fontSize = 12.sp,
                  maxLines = 1,
                  color = MaterialTheme.colors.secondary
                )
              }
            }
            Spacer(Modifier.width(8.dp))
            Icon(
              painterResource(if (selected) MR.images.ic_check_circle_filled else MR.images.ic_circle),
              contentDescription = null,
              tint = if (selected) MaterialTheme.colors.primary else MaterialTheme.colors.secondary,
              modifier = Modifier.size(24.dp)
            )
          }
        }
      }
    }
    SectionBottomSpacer()
  }
}

@Composable
private fun AddRelaysButton(onClick: () -> Unit, disabled: Boolean) {
  SettingsActionItem(
    painterResource(MR.images.ic_check),
    generalGetString(MR.strings.add_relays_title),
    click = onClick,
    textColor = MaterialTheme.colors.primary,
    iconColor = MaterialTheme.colors.primary,
    disabled = disabled,
  )
}

private suspend fun addSelectedRelays(
  groupInfo: GroupInfo,
  relayIds: List<Long>,
  selectedRelayIds: Set<Long>,
  availableRelays: List<AvailableRelay>,
  onRelayAdded: () -> Unit,
  close: () -> Unit,
  updateState: (Set<Long>, List<AvailableRelay>) -> Unit
) {
  try {
    val result = ChatController.apiAddGroupRelays(groupInfo.groupId, relayIds)
    if (result == null) {
      updateState(selectedRelayIds, availableRelays)
      return
    }
    when (result) {
      is ChatController.AddGroupRelaysResult.Added -> {
        ChannelRelaysModel.set(groupId = result.groupInfo.groupId, groupRelays = result.groupRelays)
        onRelayAdded()
        close()
      }
      is ChatController.AddGroupRelaysResult.AddFailed -> {
        val results = result.addRelayResults
        val successIds = results.filter { it.relayError == null }.mapNotNull { it.relay.chatRelayId }.toSet()
        var newSelectedIds = selectedRelayIds
        var newAvailableRelays = availableRelays
        if (successIds.isNotEmpty()) {
          newSelectedIds = selectedRelayIds - successIds
          newAvailableRelays = availableRelays.filter { it.relayId !in successIds }
          onRelayAdded()
        }
        val errorLines = results.filter { it.relayError != null }
          .map { "${chatRelayDisplayName(it.relay)}: ${it.relayError?.let { e -> ChatController.connErrorText(e) } ?: ""}" }
        val successNames = results.filter { it.relayError == null }
          .map { chatRelayDisplayName(it.relay) }
        var msg = errorLines.joinToString("\n")
        if (successNames.isNotEmpty()) {
          msg += "\n" + String.format(generalGetString(MR.strings.relays_added_format), successNames.joinToString(", "))
        }
        AlertManager.shared.showAlertMsg(
          title = generalGetString(MR.strings.error_adding_relays),
          text = msg
        )
        updateState(newSelectedIds, newAvailableRelays)
      }
    }
  } catch (e: Exception) {
    AlertManager.shared.showAlertMsg(
      title = generalGetString(MR.strings.error_adding_relays),
      text = e.message ?: ""
    )
    updateState(selectedRelayIds, availableRelays)
  }
}
